File: macos/Tests/ElevateCoreTests/ManagedProfileFetcherTests.swift
import Testing
import Foundation
@testable import ElevateCore

@Suite struct ManagedProfileFetcherTests {
    let url = URL(string: "https://contoso.example/profiles.json")!

    func temporaryCacheURL() -> URL {
        FileManager.default.temporaryDirectory
            .appendingPathComponent("elevate-tests-\(UUID().uuidString)", isDirectory: true)
            .appendingPathComponent("managed-profiles.json")
    }

    @Test func fetchParsesAndCaches() async throws {
        let cache = temporaryCacheURL()
        let http = StubHTTPClient()
        await http.on("GET", "profiles.json", body: Data(ManagedProfileSetTests.example.utf8))
        let fetcher = ManagedProfileFetcher(http: http, cacheURL: cache)

        let set = try await fetcher.fetch(from: url)
        #expect(set.profiles.map(\.id) == ["prod-incident"])
        #expect(FileManager.default.fileExists(atPath: cache.path))
        #expect(fetcher.cached() == set)

        let request = try #require(await http.requests.first)
        #expect(request.method == "GET")
        #expect(request.headers["Accept"] == "application/json")
    }

    @Test func cachedIsNilWithoutAFileOrWithAnUnreadableOne() throws {
        let cache = temporaryCacheURL()
        let fetcher = ManagedProfileFetcher(http: StubHTTPClient(), cacheURL: cache)
        #expect(fetcher.cached() == nil)

        try FileManager.default.createDirectory(at: cache.deletingLastPathComponent(), withIntermediateDirectories: true)
        try Data("not json".utf8).write(to: cache)
        #expect(fetcher.cached() == nil)
    }

    @Test func fetchThrowsOnAnErrorStatusAndLeavesTheCacheAlone() async throws {
        let cache = temporaryCacheURL()
        let http = StubHTTPClient()
        await http.on("GET", "profiles.json", status: 500, body: Data("boom".utf8))
        let fetcher = ManagedProfileFetcher(http: http, cacheURL: cache)

        await #expect(throws: ManagedProfileError.invalid("managed profiles fetch failed with HTTP 500")) {
            try await fetcher.fetch(from: url)
        }
        #expect(!FileManager.default.fileExists(atPath: cache.path))
    }

    @Test func fetchThrowsOnAnOversizedBody() async throws {
        let cache = temporaryCacheURL()
        let http = StubHTTPClient()
        let big = Data(repeating: UInt8(ascii: " "), count: ManagedProfileFetcher.maxBytes + 1)
        await http.on("GET", "profiles.json", body: big)
        let fetcher = ManagedProfileFetcher(http: http, cacheURL: cache)

        await #expect(throws: ManagedProfileError.invalid("managed profiles document is larger than 1 MB")) {
            try await fetcher.fetch(from: url)
        }
        #expect(!FileManager.default.fileExists(atPath: cache.path))
    }

    @Test func fetchRefusesANonHttpsURL() async throws {
        let fetcher = ManagedProfileFetcher(http: StubHTTPClient(), cacheURL: temporaryCacheURL())
        await #expect(throws: ManagedProfileError.invalid("managed profiles URL must be https")) {
            try await fetcher.fetch(from: URL(string: "http://contoso.example/profiles.json")!)
        }
    }

    @Test func fetchThrowsOnAnUnparsableBody() async throws {
        let cache = temporaryCacheURL()
        let http = StubHTTPClient()
        await http.on("GET", "profiles.json", body: Data(#"{"version":2}"#.utf8))
        let fetcher = ManagedProfileFetcher(http: http, cacheURL: cache)

        await #expect(throws: ManagedProfileError.invalid("version 2 is not supported")) {
            try await fetcher.fetch(from: url)
        }
        #expect(!FileManager.default.fileExists(atPath: cache.path))
    }
}

File: windows/tests/Elevate.Core.Tests/Managed/ManagedProfileFetcherTests.cs
using System.Text;
using Elevate.Core.Managed;
using Elevate.Core.Tests.Support;
using FluentAssertions;

namespace Elevate.Core.Tests.Managed;

public class ManagedProfileFetcherTests
{
    private static readonly Uri Url = new("https://contoso.example/profiles.json");

    private static string TemporaryCachePath()
        => Path.Combine(Path.GetTempPath(), "elevate-tests-" + Guid.NewGuid().ToString("N"), "managed-profiles.json");

    [Fact]
    public async Task FetchParsesAndCaches()
    {
        var cache = TemporaryCachePath();
        var http = new StubHttpClient();
        http.On("GET", "profiles.json", ManagedProfileSetTests.Example);
        var fetcher = new ManagedProfileFetcher(http, cache);

        var set = await fetcher.FetchAsync(Url);

        set.Profiles.Select(p => p.Id).Should().Equal("prod-incident");
        File.Exists(cache).Should().BeTrue();
        fetcher.Cached().Should().Be(set);

        var request = http.Requests.Should().ContainSingle().Subject;
        request.Method.Should().Be("GET");
        request.Headers["Accept"].Should().Be("application/json");
    }

    [Fact]
    public void CachedIsNullWithoutAFileOrWithAnUnreadableOne()
    {
        var cache = TemporaryCachePath();
        var fetcher = new ManagedProfileFetcher(new StubHttpClient(), cache);

        fetcher.Cached().Should().BeNull();

        Directory.CreateDirectory(Path.GetDirectoryName(cache)!);
        File.WriteAllText(cache, "not json");
        fetcher.Cached().Should().BeNull();
    }

    [Fact]
    public async Task FetchThrowsOnAnErrorStatusAndLeavesTheCacheAlone()
    {
        var cache = TemporaryCachePath();
        var http = new StubHttpClient();
        http.On("GET", "profiles.json", "boom", status: 500);
        var fetcher = new ManagedProfileFetcher(http, cache);

        var act = () => fetcher.FetchAsync(Url);

        (await act.Should().ThrowAsync<ManagedProfileException>()).Which.Message
            .Should().Be("managed profiles fetch failed with HTTP 500");
        File.Exists(cache).Should().BeFalse();
    }

    [Fact]
    public async Task FetchThrowsOnAnOversizedBody()
    {
        var cache = TemporaryCachePath();
        var http = new StubHttpClient();
        http.On("GET", "profiles.json", body: Encoding.UTF8.GetBytes(new string(' ', ManagedProfileFetcher.MaxBytes + 1)));
        var fetcher = new ManagedProfileFetcher(http, cache);

        var act = () => fetcher.FetchAsync(Url);

        (await act.Should().ThrowAsync<ManagedProfileException>()).Which.Message
            .Should().Be("managed profiles document is larger than 1 MB");
        File.Exists(cache).Should().BeFalse();
    }

    [Fact]
    public async Task FetchRefusesANonHttpsUrl()
    {
        var fetcher = new ManagedProfileFetcher(new StubHttpClient(), TemporaryCachePath());

        var act = () => fetcher.FetchAsync(new Uri("http://contoso.example/profiles.json"));

        (await act.Should().ThrowAsync<ManagedProfileException>()).Which.Message
            .Should().Be("managed profiles URL must be https");
    }

    [Fact]
    public async Task FetchThrowsOnAnUnparsableBody()
    {
        var cache = TemporaryCachePath();
        var http = new StubHttpClient();
        http.On("GET", "profiles.json", """{"version":2}""");
        var fetcher = new ManagedProfileFetcher(http, cache);

        var act = () => fetcher.FetchAsync(Url);

        (await act.Should().ThrowAsync<ManagedProfileException>()).Which.Message.Should().Be("version 2 is not supported");
        File.Exists(cache).Should().BeFalse();
    }
}
